src/hotspot/share/gc/shenandoah/shenandoahLock.hpp

```diff
@@ -32,40 +32,43 @@
 
 class ShenandoahLock  {
 private:
-  enum LockState { unlocked = 0, locked = 1 };
-
   shenandoah_padding(0);
   volatile int _state;
   shenandoah_padding(1);
   volatile Thread* _owner;
   shenandoah_padding(2);
 
+  void contended_lock_or_block(JavaThread* java_thread);
+  void contended_lock_no_block();
+
 public:
-  ShenandoahLock() : _state(unlocked), _owner(nullptr) {};
+  ShenandoahLock() : _state(0), _owner(nullptr) {};
 
-  void lock() {
-#ifdef ASSERT
+  void lock(bool allow_block_for_safepoint) {
     assert(_owner != Thread::current(), "reentrant locking attempt, would deadlock");
-#endif
-    Thread::SpinAcquire(&_state, "Shenandoah Heap Lock");
-#ifdef ASSERT
-    assert(_state == locked, "must be locked");
+
+    // Try to lock fast, or dive into contended lock handling.
+    if (Atomic::cmpxchg(&_state, 0, 1) != 0) {
```

Comment:
If we keep enum, this line reads:
`if (Atomic::cmpxchg(&_state, unlocked, locked) != unlocked) {`

Which I find more appealing.